lib/Backends/Backend.cpp

```diff
@@ -15,23 +15,75 @@
 
-void Backend::optimizeFunction(Function *F,
-                               const CompilationOptions &opts) const {
+static void transformForPrecisionMode(const Backend &B, Function *&F,
+                                      CompilationOptions &opts) {
+  if (opts.profileForQuantization) {
+    assert(opts.bindings);
+    Function *PF = glow::profileQuantization(*opts.bindings, F);
+    // Erase the original function so that the redundant variables that are only
+    // referenced by the original function will be removed.
+    F->getParent()->eraseFunction(F);
+    F = PF;
+  }
+
+  if (opts.quantize) {
+    // In AOT compilation mode the name of the symbol depends on the name of the
+    // function. Our tutorial expects the quantized name to be identical to the
+    // original name, so we rename the floating-point network and give the
+    // quantized network a new name.
+    std::string oldName = F->getName();
+    F->setName("old");
+
+    Function *QF = quantization::quantizeFunction(
+        B, opts.quantizationSchema, opts.quantizationInfos,
+        opts.quantizationPrecision, F, opts.loweredInfoMap, oldName,
+        opts.precisionModeKindSet, opts.enableRowwiseQuantization);
+
+    // Erase the original function so that the redundant variables that are only
+    // referenced by the original function will be removed.
+    F->getParent()->eraseFunction(F);
+    F = QF;
+  }
+
+  if (opts.convertToFP16) {
+    TypeAToTypeBFunctionConverter converter(
+        *F, ElemKind::FloatTy, ElemKind::Float16Ty, &opts.precisionModeKindSet);
+    converter.convert();
+  }
+}
+
+void Backend::optimizeFunction(Function *&F, CompilationOptions &opts) const {
   // Verify the function pre-optimization/lowering.
   assert(F->verify() && "Function must be valid");
 
+  // Fold low-level operators into higher-level operators.
+  // This is useful when compiling an input model where some high-level
+  // operators have been lowered (this can be for instance a side effect of
+  // model converters, like converters from Tensorflow to ONNX). In this
+  // situation, such folding can then enable more optimizations and also improve
+  // the performance backends that support natively such high-level operators.
+  ::glow::fold(F, glow::CompilationMode::Infer);
+
   // Optimize the graph.
   ::glow::optimize(F, opts);
 
-  // Lower the graph into a sequence of low-level linear algebra operations.
-  ::glow::lower(F, /* loweredMap */ nullptr, this);
+  // Lower the graph into a sequence of low-level linear algebra operations. If
+  // profiling, pass a nullptr for the backend, as all nodes should be lowered.
+  ::glow::lower(F, &opts.loweredInfoMap,
+                opts.profileForQuantization ? nullptr : this,
+                opts.precisionModeKindSet);
+
+  // Transform given precision mode; may quantize, convert to fp16, or
+  // instrument with profiling nodes. This must be done after lowering.
+  transformForPrecisionMode(*this, F, opts);
```

Comment:
Should transformForPrecisionMode be moved to global scope like the rest of the function here, there's no real reason to have it be in Backend.cpp right?

Reply:
I guess if we want to unit test it then making it global would make sense; otherwise it's basically just a helper that was extracted to make the logic in `optimizeFunction()` more readable.

Additionally, I think it along with `optimizeFunction()` should be moved to GraphOptimizer.cpp (see https://github.com/pytorch/glow/pull/2703).